File: main/contrib/Windows API Code Pack 1.1/source/Samples/ExplorerBrowser/VB/WinForms/ExplorerBrowserTestForm.vb
'Copyright (c) Microsoft Corporation.  All rights reserved.


Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports Microsoft.WindowsAPICodePack.Shell
Imports System.Windows.Forms.Integration
Imports System.Reflection
Imports System.Runtime.InteropServices
Imports Microsoft.WindowsAPICodePack.Controls
Imports System.Threading

Namespace Microsoft.WindowsAPICodePack.Samples
	Partial Public Class ExplorerBrowserTestForm
        Inherits Form
        Dim uiDecoupleTimer As New System.Windows.Forms.Timer
        Dim selectionChanged As New AutoResetEvent(False)
        Dim itemsChanged As New AutoResetEvent(False)

		Public Sub New()
			InitializeComponent()

			' initialize known folder combo box
			Dim knownFolderList As New List(Of String)()
			For Each folder As IKnownFolder In KnownFolders.All
				knownFolderList.Add(folder.CanonicalName)
			Next folder
			knownFolderList.Sort()
			knownFolderCombo.Items.AddRange(knownFolderList.ToArray())

			' initial property grids
			navigationPropertyGrid.SelectedObject = explorerBrowser.NavigationOptions
			visibilityPropertyGrid.SelectedObject = explorerBrowser.NavigationOptions.PaneVisibility
			contentPropertyGrid.SelectedObject = explorerBrowser.ContentOptions

			' setup ExplorerBrowser navigation events
			AddHandler explorerBrowser.NavigationPending, AddressOf explorerBrowser_NavigationPending
			AddHandler explorerBrowser.NavigationFailed, AddressOf explorerBrowser_NavigationFailed
			AddHandler explorerBrowser.NavigationComplete, AddressOf explorerBrowser_NavigationComplete
			AddHandler explorerBrowser.ItemsChanged, AddressOf explorerBrowser_ItemsChanged
            AddHandler explorerBrowser.SelectionChanged, AddressOf explorerBrowser_SelectionChanged
            AddHandler explorerBrowser.ViewEnumerationComplete, AddressOf explorerBrowser_ViewEnumerationComplete


			' set up Navigation log event and button state
			AddHandler explorerBrowser.NavigationLog.NavigationLogChanged, AddressOf NavigationLog_NavigationLogChanged
			Me.backButton.Enabled = False
            Me.forwardButton.Enabled = False

            AddHandler uiDecoupleTimer.Tick, AddressOf uiDecoupleTimer_Tick
            Me.uiDecoupleTimer.Interval = 100
            uiDecoupleTimer.Start()

		End Sub

        Private Sub uiDecoupleTimer_Tick( ByVal sender As Object, ByVal args As EventArgs )
            If (selectionChanged.WaitOne(1)) Then
                Dim itemsText As New StringBuilder()
                For Each item As ShellObject In explorerBrowser.SelectedItems
                    If item IsNot Nothing Then
                        itemsText.AppendLine(Constants.vbTab & "Item = " & item.GetDisplayName(DisplayNameType.Default))
                    End If
                Next item
                Me.selectedItemsTextBox.Text = itemsText.ToString()
                Me.itemsTabControl.TabPages(1).Text = "Selected Items (Count=" & explorerBrowser.SelectedItems.Count.ToString() & ")"
            End If

            If (itemsChanged.WaitOne(1)) Then
                Dim itemsText As New StringBuilder()
                For Each item As ShellObject In explorerBrowser.Items
                    If item IsNot Nothing Then
                        itemsText.AppendLine(Constants.vbTab & "Item = " & item.GetDisplayName(DisplayNameType.Default))
                    End If
                Next item
                Me.itemsTextBox.Text = itemsText.ToString()
                Me.itemsTabControl.TabPages(0).Text = "Items (Count=" & explorerBrowser.Items.Count.ToString() & ")"
            End If
        End Sub

        Private Sub explorerBrowser_ViewEnumerationComplete(ByVal sender As Object, ByVal args As EventArgs)
            ' This event is BeginInvoked to decouple the ExplorerBrowser UI from this UI
            BeginInvoke(New MethodInvoker(Function() AnonymousMethod2()))

            selectionChanged.Set()
            itemsChanged.Set()
        End Sub
        Private Function AnonymousMethod2() As Object
            Me.eventHistoryTextBox.Text = Me.eventHistoryTextBox.Text & "View enumeration complete." & Constants.vbLf
            Return Nothing
        End Function


        Protected Overrides Sub OnShown(ByVal e As EventArgs)
            MyBase.OnShown(e)
            explorerBrowser.Navigate(CType(KnownFolders.Desktop, ShellObject))
        End Sub

        Private Sub NavigationLog_NavigationLogChanged(ByVal sender As Object, ByVal args As NavigationLogEventArgs)
            ' This event is BeginInvoked to decouple the ExplorerBrowser UI from this UI
            ' calculate button states
            ' update history combo box
            BeginInvoke(New MethodInvoker(Function() AnonymousMethod1(args)))
        End Sub

        Private Function AnonymousMethod1(ByVal args As NavigationLogEventArgs) As Object
            If args.CanNavigateBackwardChanged Then
                Me.backButton.Enabled = explorerBrowser.NavigationLog.CanNavigateBackward
            End If
            If args.CanNavigateForwardChanged Then
                Me.forwardButton.Enabled = explorerBrowser.NavigationLog.CanNavigateForward
            End If
            If args.LocationsChanged Then
                Me.navigationHistoryCombo.Items.Clear()
                For Each shobj As ShellObject In Me.explorerBrowser.NavigationLog.Locations
                    Me.navigationHistoryCombo.Items.Add(shobj.Name)
                Next shobj
            End If
            If Me.explorerBrowser.NavigationLog.CurrentLocationIndex = -1 Then
                Me.navigationHistoryCombo.Text = ""
            Else
                Me.navigationHistoryCombo.SelectedIndex = Me.explorerBrowser.NavigationLog.CurrentLocationIndex
            End If
            Return Nothing
        End Function

        Private Sub explorerBrowser_SelectionChanged(ByVal sender As Object, ByVal e As EventArgs)
            selectionChanged.Set()
        End Sub

        Private Sub explorerBrowser_ItemsChanged(ByVal sender As Object, ByVal e As EventArgs)
            itemsChanged.Set()
        End Sub

        Private Sub explorerBrowser_NavigationComplete(ByVal sender As Object, ByVal args As NavigationCompleteEventArgs)
            ' This event is BeginInvoked to decouple the ExplorerBrowser UI from this UI
            ' update event history text box
            BeginInvoke(New MethodInvoker(Function() AnonymousMethod4(args)))
        End Sub

        Private Function AnonymousMethod4(ByVal args As NavigationCompleteEventArgs) As Object
            Dim location As String = If((args.NewLocation Is Nothing), "(unknown)", args.NewLocation.Name)
            Me.eventHistoryTextBox.Text = Me.eventHistoryTextBox.Text & "Navigation completed. New Location = " & location & Constants.vbLf
            Return Nothing
        End Function

        Private Sub explorerBrowser_NavigationFailed(ByVal sender As Object, ByVal args As NavigationFailedEventArgs)
            ' This event is BeginInvoked to decouple the ExplorerBrowser UI from this UI
            ' update event history text box
            BeginInvoke(New MethodInvoker(Function() AnonymousMethod5(args)))
        End Sub

        Private Function AnonymousMethod5(ByVal args As NavigationFailedEventArgs) As Object
            Dim location As String = If((args.FailedLocation Is Nothing), "(unknown)", args.FailedLocation.Name)
            Me.eventHistoryTextBox.Text = Me.eventHistoryTextBox.Text & "Navigation failed. Failed Location = " & location & Constants.vbLf
            If Me.explorerBrowser.NavigationLog.CurrentLocationIndex = -1 Then
                Me.navigationHistoryCombo.Text = ""
            Else
                Me.navigationHistoryCombo.SelectedIndex = Me.explorerBrowser.NavigationLog.CurrentLocationIndex
            End If
            Return Nothing
        End Function

        Private Sub explorerBrowser_NavigationPending(ByVal sender As Object, ByVal args As NavigationPendingEventArgs)
            ' fail navigation if check selected (this must be synchronous)
            args.Cancel = Me.failNavigationCheckBox.Checked


            ' This portion is BeginInvoked to decouple the ExplorerBrowser UI from this UI
            ' update event history text box
            BeginInvoke(New MethodInvoker(Function() AnonymousMethod6(args)))
        End Sub

        Private Function AnonymousMethod6(ByVal args As NavigationPendingEventArgs) As Object
            Dim message As String = ""
            Dim location As String = If((args.PendingLocation Is Nothing), "(unknown)", args.PendingLocation.Name)
            If args.Cancel Then
                message = "Navigation Failing. Pending Location = " & location
            Else
                message = "Navigation Pending. Pending Location = " & location
            End If
            Me.eventHistoryTextBox.Text = Me.eventHistoryTextBox.Text & message & Constants.vbLf
            Return Nothing
        End Function

        Private Sub navigateButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles navigateButton.Click
            Try
                ' navigate to specific folder
                explorerBrowser.Navigate(ShellFileSystemFolder.FromFolderPath(pathEdit.Text))
            Catch e1 As COMException
                MessageBox.Show("Navigation not possible.")
            End Try
        End Sub

        Private Sub filePathNavigate_Click(ByVal sender As Object, ByVal e As EventArgs) Handles filePathNavigate.Click
            Try
                ' Navigates to a specified file (must be a container file to work, i.e., ZIP, CAB)         
                Me.explorerBrowser.Navigate(ShellFile.FromFilePath(Me.filePathEdit.Text))
            Catch e1 As COMException
                MessageBox.Show("Navigation not possible.")
            End Try
        End Sub

        Private Sub knownFolderNavigate_Click(ByVal sender As Object, ByVal e As EventArgs) Handles knownFolderNavigate.Click
            Try
                ' Navigate to a known folder
                Dim kf As IKnownFolder = KnownFolderHelper.FromCanonicalName(Me.knownFolderCombo.Items(knownFolderCombo.SelectedIndex).ToString())

                Me.explorerBrowser.Navigate(CType(kf, ShellObject))
            Catch e1 As COMException
                MessageBox.Show("Navigation not possible.")
            End Try
        End Sub

        Private Sub forwardButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles forwardButton.Click
            ' Move forwards through navigation log
            explorerBrowser.NavigateLogLocation(NavigationLogDirection.Forward)
        End Sub

        Private Sub backButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles backButton.Click
            ' Move backwards through navigation log
            explorerBrowser.NavigateLogLocation(NavigationLogDirection.Backward)
        End Sub

        Private Sub navigationHistoryCombo_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles navigationHistoryCombo.SelectedIndexChanged
            ' navigating to specific index in navigation log
            explorerBrowser.NavigateLogLocation(Me.navigationHistoryCombo.SelectedIndex)
        End Sub

        Private Sub clearHistoryButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles clearHistoryButton.Click
            ' clear navigation log
            explorerBrowser.NavigationLog.ClearLog()
        End Sub

        Private Sub filePathEdit_TextChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles filePathEdit.TextChanged
            filePathNavigate.Enabled = (filePathEdit.Text.Length > 0)
        End Sub

        Private Sub pathEdit_TextChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles pathEdit.TextChanged
            navigateButton.Enabled = (pathEdit.Text.Length > 0)
        End Sub

        Private Sub knownFolderCombo_TextChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles knownFolderCombo.TextChanged
            knownFolderNavigate.Enabled = (knownFolderCombo.Text.Length > 0)
        End Sub

    End Class
End Namespace

File: main/contrib/Windows API Code Pack 1.1/source/Samples/ExplorerBrowser/CS/WinForms/ExplorerBrowserTestForm.cs
//Copyright (c) Microsoft Corporation.  All rights reserved.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using Microsoft.WindowsAPICodePack.Shell;
using System.Windows.Forms.Integration;
using System.Reflection;
using System.Runtime.InteropServices;
using Microsoft.WindowsAPICodePack.Controls;
using System.Threading;

namespace Microsoft.WindowsAPICodePack.Samples
{
    public partial class ExplorerBrowserTestForm : Form
    {
        System.Windows.Forms.Timer uiDecoupleTimer = new System.Windows.Forms.Timer();
        AutoResetEvent selectionChanged = new AutoResetEvent(false);
        AutoResetEvent itemsChanged = new AutoResetEvent(false);

        public ExplorerBrowserTestForm()
        {
            InitializeComponent();

            // initialize known folder combo box
            List<string> knownFolderList = new List<string>();
            foreach (IKnownFolder folder in KnownFolders.All)
            {
                knownFolderList.Add(folder.CanonicalName);
            }
            knownFolderList.Sort();
            knownFolderCombo.Items.AddRange(knownFolderList.ToArray());

            // initial property grids
            navigationPropertyGrid.SelectedObject = explorerBrowser.NavigationOptions;
            visibilityPropertyGrid.SelectedObject = explorerBrowser.NavigationOptions.PaneVisibility;
            contentPropertyGrid.SelectedObject = explorerBrowser.ContentOptions;

            // setup ExplorerBrowser navigation events
            explorerBrowser.NavigationPending += new EventHandler<NavigationPendingEventArgs>(explorerBrowser_NavigationPending);
            explorerBrowser.NavigationFailed += new EventHandler<NavigationFailedEventArgs>(explorerBrowser_NavigationFailed);
            explorerBrowser.NavigationComplete += new EventHandler<NavigationCompleteEventArgs>(explorerBrowser_NavigationComplete);
            explorerBrowser.ItemsChanged += new EventHandler(explorerBrowser_ItemsChanged);
            explorerBrowser.SelectionChanged += new EventHandler(explorerBrowser_SelectionChanged);
            explorerBrowser.ViewEnumerationComplete += new EventHandler(explorerBrowser_ViewEnumerationComplete);

            // set up Navigation log event and button state
            explorerBrowser.NavigationLog.NavigationLogChanged += new EventHandler<NavigationLogEventArgs>(NavigationLog_NavigationLogChanged);
            this.backButton.Enabled = false;
            this.forwardButton.Enabled = false;

            uiDecoupleTimer.Tick += new EventHandler(uiDecoupleTimer_Tick);
            uiDecoupleTimer.Interval = 100;
            uiDecoupleTimer.Start();
        }

        void uiDecoupleTimer_Tick(object sender, EventArgs e)
        {
            if (selectionChanged.WaitOne(1))
            {
                StringBuilder itemsText = new StringBuilder();

                foreach (ShellObject item in explorerBrowser.SelectedItems)
                {
                    if (item != null)
                        itemsText.AppendLine("\tItem = " + item.GetDisplayName(DisplayNameType.Default));
                }

                this.selectedItemsTextBox.Text = itemsText.ToString();
                this.itemsTabControl.TabPages[1].Text = "Selected Items (Count=" + explorerBrowser.SelectedItems.Count.ToString() + ")";
            }

            if (itemsChanged.WaitOne(1))
            {
                // update items text box
                StringBuilder itemsText = new StringBuilder();

                foreach (ShellObject item in explorerBrowser.Items)
                {
                    if (item != null)
                        itemsText.AppendLine("\tItem = " + item.GetDisplayName(DisplayNameType.Default));
                }

                this.itemsTextBox.Text = itemsText.ToString();
                this.itemsTabControl.TabPages[0].Text = "Items (Count=" + explorerBrowser.Items.Count.ToString() + ")";
            }
        }

        void explorerBrowser_ViewEnumerationComplete(object sender, EventArgs e)
        {
            // This event is BeginInvoked to decouple the ExplorerBrowser UI from this UI
            BeginInvoke(new MethodInvoker(delegate()
            {
                this.eventHistoryTextBox.Text =
                    this.eventHistoryTextBox.Text +
                    "View enumeration complete.\n";
            }));

            selectionChanged.Set();
            itemsChanged.Set();
        }


        protected override void OnShown(EventArgs e)
        {
            base.OnShown(e);
            explorerBrowser.Navigate((ShellObject)KnownFolders.Desktop);
        }

        void NavigationLog_NavigationLogChanged(object sender, NavigationLogEventArgs args)
        {
            // This event is BeginInvoked to decouple the ExplorerBrowser UI from this UI
            BeginInvoke(new MethodInvoker(delegate()
            {
                // calculate button states
                if (args.CanNavigateBackwardChanged)
                {
                    this.backButton.Enabled = explorerBrowser.NavigationLog.CanNavigateBackward;
                }
                if (args.CanNavigateForwardChanged)
                {
                    this.forwardButton.Enabled = explorerBrowser.NavigationLog.CanNavigateForward;
                }

                // update history combo box
                if (args.LocationsChanged)
                {
                    this.navigationHistoryCombo.Items.Clear();
                    foreach (ShellObject shobj in this.explorerBrowser.NavigationLog.Locations)
                    {
                        this.navigationHistoryCombo.Items.Add(shobj.Name);
                    }
                }
                if (this.explorerBrowser.NavigationLog.CurrentLocationIndex == -1)
                    this.navigationHistoryCombo.Text = "";
                else
                    this.navigationHistoryCombo.SelectedIndex = this.explorerBrowser.NavigationLog.CurrentLocationIndex;
            }));
        }

        void explorerBrowser_SelectionChanged(object sender, EventArgs e)
        {
            selectionChanged.Set();
        }

        void explorerBrowser_ItemsChanged(object sender, EventArgs e)
        {
            itemsChanged.Set();
        }

        void explorerBrowser_NavigationComplete(object sender, NavigationCompleteEventArgs args)
        {
            // This event is BeginInvoked to decouple the ExplorerBrowser UI from this UI
            BeginInvoke(new MethodInvoker(delegate()
            {
                // update event history text box
                string location = (args.NewLocation == null) ? "(unknown)" : args.NewLocation.Name;
                this.eventHistoryTextBox.Text =
                    this.eventHistoryTextBox.Text +
                    "Navigation completed. New Location = " + location + "\n";
            }));
        }

        void explorerBrowser_NavigationFailed(object sender, NavigationFailedEventArgs args)
        {
            // This event is BeginInvoked to decouple the ExplorerBrowser UI from this UI
            BeginInvoke(new MethodInvoker(delegate()
            {
                // update event history text box
                string location = (args.FailedLocation == null) ? "(unknown)" : args.FailedLocation.Name;
                this.eventHistoryTextBox.Text =
                    this.eventHistoryTextBox.Text +
                    "Navigation failed. Failed Location = " + location + "\n";

                if (this.explorerBrowser.NavigationLog.CurrentLocationIndex == -1)
                    this.navigationHistoryCombo.Text = "";
                else
                    this.navigationHistoryCombo.SelectedIndex = this.explorerBrowser.NavigationLog.CurrentLocationIndex;
            }));
        }

        void explorerBrowser_NavigationPending(object sender, NavigationPendingEventArgs args)
        {
            // fail navigation if check selected (this must be synchronous)
            args.Cancel = this.failNavigationCheckBox.Checked;


            // This portion is BeginInvoked to decouple the ExplorerBrowser UI from this UI
            BeginInvoke(new MethodInvoker(delegate()
            {
                // update event history text box
                string message = "";
                string location = (args.PendingLocation == null) ? "(unknown)" : args.PendingLocation.Name;
                if (args.Cancel)
                {
                    message = "Navigation Failing. Pending Location = " + location;
                }
                else
                {
                    message = "Navigation Pending. Pending Location = " + location;
                }
                this.eventHistoryTextBox.Text =
                    this.eventHistoryTextBox.Text + message + "\n";
            }));
        }

        private void navigateButton_Click(object sender, EventArgs e)
        {
            try
            {
                // navigate to specific folder
                explorerBrowser.Navigate(ShellFileSystemFolder.FromFolderPath(pathEdit.Text));
            }
            catch (COMException)
            {
                MessageBox.Show("Navigation not possible.");
            }
        }

        private void filePathNavigate_Click(object sender, EventArgs e)
        {
            try
            {
                // Navigates to a specified file (must be a container file to work, i.e., ZIP, CAB)         
                this.explorerBrowser.Navigate(ShellFile.FromFilePath(this.filePathEdit.Text));
            }
            catch (COMException)
            {
                MessageBox.Show("Navigation not possible.");
            }
        }

        private void knownFolderNavigate_Click(object sender, EventArgs e)
        {
            try
            {
                // Navigate to a known folder
                IKnownFolder kf =
                    KnownFolderHelper.FromCanonicalName(
                        this.knownFolderCombo.Items[knownFolderCombo.SelectedIndex].ToString());

                this.explorerBrowser.Navigate((ShellObject)kf);
            }
            catch (COMException)
            {
                MessageBox.Show("Navigation not possible.");
            }
        }

        private void forwardButton_Click(object sender, EventArgs e)
        {
            // Move forwards through navigation log
            explorerBrowser.NavigateLogLocation(NavigationLogDirection.Forward);
        }

        private void backButton_Click(object sender, EventArgs e)
        {
            // Move backwards through navigation log
            explorerBrowser.NavigateLogLocation(NavigationLogDirection.Backward);
        }

        private void navigationHistoryCombo_SelectedIndexChanged(object sender, EventArgs e)
        {
            // navigating to specific index in navigation log
            explorerBrowser.NavigateLogLocation(this.navigationHistoryCombo.SelectedIndex);
        }

        private void clearHistoryButton_Click(object sender, EventArgs e)
        {
            // clear navigation log
            explorerBrowser.NavigationLog.ClearLog();
        }

        private void filePathEdit_TextChanged(object sender, EventArgs e)
        {
            filePathNavigate.Enabled = (filePathEdit.Text.Length > 0);
        }

        private void pathEdit_TextChanged(object sender, EventArgs e)
        {
            navigateButton.Enabled = (pathEdit.Text.Length > 0);
        }

        private void knownFolderCombo_SelectedIndexChanged(object sender, EventArgs e)
        {
            knownFolderNavigate.Enabled = (knownFolderCombo.Text.Length > 0);
        }
    }
}
